lib/runner/ssh-tunnel-agent.js

```diff
@@ -0,0 +1,221 @@
+        var self = this,
+            host = sshConfig.auth.host;
+
+        sshClient.forwardOut('', sock.localPort, sshConfig.auth.host, sshConfig.proxyPort, function(err, stream) {
+            if (err) {
+                self.reporter.emit('tunnelError', new Error('request cannot be tunneled to host: ' + host +
+                        ', ' + err.message));
+                self.closeSock(sock, stream);
+
+                return;
+            }
+
+            sock.pipe(stream).pipe(sock);
+
+            sock.on('error', function(err) {
+                self.reporter.emit('tunnelError', new Error('connection was closed with errors for tunnel host: ' +
+                        host + ', ' + err.message));
+                self.closeSock(sock, stream);
+            });
+
+            stream.on('error', function(err) {
+                self.reporter.emit('tunnelError', new Error('connection was closed with errors for tunnel host: ' +
+                        host + ', ') + err.message);
+                self.closeSock(sock, stream);
+            });
+        });
+    },
+
+    /**
+     * Given a proxy list creates and starts different ssh tunnels based on proxy list.
+     *
+     * @param {ProxyConfigList} proxyConfigList list of proxies
+     * @param {Number} localport port listening locally for requests and forwarding them through tunnel
+     * @param {Object} scope
+     * @param {Function} cb callback function that will be called once all the tunnels are
+     * initialised or some error occurred
+     *
+     */
+
+    start: function (proxyConfigList, localport, scope, cb) {
+        var self = this,
+            sshConfigList;
+
+
+        self.proxyConfigList = (ProxyConfigList.isProxyConfigList(proxyConfigList)) ?
+            proxyConfigList : (_.isArray(proxyConfigList) && new ProxyConfigList({}, proxyConfigList));
+
+        sshConfigList = self.proxyConfigList.map(function (proxyConfig) {
+            return proxyConfig.sshTunnelConfig;
+        }, self);
+
+        // locally listen on a port, if provided port value is not valid pass 0 and a random pprt will be assigned
+        (_.isInteger(localport) && (localport >= 0) && (localport <= HIGHEST_POSSIBLE_PORT)) ?
+            localport : (localport = DEFAULT_LOCAL_PORT);
+
+        self.setLocalProxy(scope, localport);
+
+        self.localServer = net.createServer(function(sock) {
+            // resolve proxyconfig to be used based on the request url
+            var url = scope.currentReq,
+                client,
+                sshConfig,
+                error,
+                pc = self.proxyConfigList.resolve(url);
+
+            // if proxy cannot be resolved
+            if (pc) {
+                sshConfig = _.filter((_.uniq(self.mapping)), ['config', pc.sshTunnelConfig]);
+                client = _.isEmpty(sshConfig) ? false : _.head(sshConfig).client;
+                error = new Error('Can not resolve proxy for: ' + url);
```

Comment:
Call `new Error` only if there's an actual error, this is confusing to read.

Reply:
it is an actual new error